src/Http/WebUtilities/src/FileBufferingWriteStream.cs

```diff
@@ -0,0 +1,308 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Internal;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    /// <summary>
+    /// A <see cref="Stream"/> that wraps another stream and buffers content to be written.
+    /// <para>
+    /// <see cref="FileBufferingWriteStream"/> is intended to provide a workaround for APIs that need to perform
+    /// synchronous writes to the HTTP Response Stream while contending with a server that is configured to disallow synchronous writes.
+    /// Synchronous writes are buffered to a memory backed stream up to the specified threshold, after which further writes are spooled to
+    /// a temporary file on disk.
+    /// </para>
+    /// <para>
+    /// The <see cref="FileBufferingWriteStream"/> performs opportunistic writes to the wrapping stream
+    /// when asychronous operation such as <see cref="WriteAsync(byte[], int, int, CancellationToken)"/> or <see cref="FlushAsync(CancellationToken)"/>
+    /// are performed.
+    /// </para>
+    /// </summary>
+    public sealed class FileBufferingWriteStream : Stream
+    {
+        private const int MaxRentedBufferSize = 1024 * 1024; // 1MB
+        private const int DefaultMemoryThreshold = 30 * 1024; // 30k
+
+        private readonly Stream _writeStream;
+        private readonly int _memoryThreshold;
+        private readonly long? _bufferLimit;
+        private readonly Func<string> _tempFileDirectoryAccessor;
+        private readonly ArrayPool<byte> _bytePool;
+        private readonly byte[] _rentedBuffer;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="FileBufferingWriteStream"/>.
+        /// </summary>
+        /// <param name="writeStream">The <see cref="Stream"/> to write buffered contents to.</param>
+        /// <param name="memoryThreshold">
+        /// The maximum amount of memory in bytes to allocate before switching to a file on disk.
+        /// Defaults to 30kb.
+        /// </param>
+        /// <param name="bufferLimit">
+        /// The maximum amouont of bytes that the <see cref="FileBufferingWriteStream"/> is allowed to buffer.
+        /// </param>
+        /// <param name="tempFileDirectoryAccessor">Provides the location of the directory to write buffered contents to.
+        /// When unspecified, uses the value specified by the environment variable <c>ASPNETCORE_TEMP</c> if available, otherwise
+        /// uses the value returned by <see cref="Path.GetTempPath"/>.
+        /// </param>
+        public FileBufferingWriteStream(
+            Stream writeStream,
```

Comment:
Is there a reason why we wrap another stream instead of writing all of the data and then doing a copy to async? The problem with this pattern (abstracting behind a stream) is that it only works if you call `FlushAsync` on the stream - so the abstraction leaks.

Basically it seems like you have to know about buffering to do this properly, which gets me back to why we wrap the stream instead of writing everything and then copying.